### PR TITLE
CARBON-968 - Adding Portal component to input-validation

### PR DESCRIPTION
**Pull request by whomwah**
At the moment I just want to get some feedback as this was quite a big change, as the inclusion of `Portal` broken quite a large part of the validation/state functionality which I am slowly getting back.

# TODO
- [x] Specs and coverage
- [x] Release notes

# Related Issues / Pull Requests

Jira Ticket: CARBON-968

# Screenshots / Gifs
scroll

# Testing Instructions

### CODE REVIEW

File: src/utils/decorators/input-validation/input-validation.js [hunk 1]
```diff
@@ -167,6 +171,93 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
     info: PropTypes.array
   });
 
+  componentDidUpdate(prevProps) {
+    if (super.componentDidUpdate) { super.componentDidUpdate(prevProps); }
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: this function also receives `prevState`

File: src/utils/decorators/input-validation/input-validation.js [hunk 2]
```diff
@@ -167,6 +171,93 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
...
+
+    if (!this.state.valid || this.state.warning || this.state.info) {
+      this.positionElements();
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: do we need to run this on every update? can we try and minimise how many times it needs to be called somehow?

File: src/utils/decorators/input-validation/input-validation.js [hunk 3]
```diff
@@ -227,6 +318,18 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
   }
 
   /**
+   * A lifecycle method for when the component has mounted.
+   *
+   * @method componentDidMount
+   * @return {void}
+   */
+  componentDidMount() {
+    if (typeof this._window !== 'undefined') {
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: does this need to call super too?

File: src/utils/decorators/input-validation/input-validation.js [hunk 4]
```diff
@@ -167,6 +171,93 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
...
+    return ReactDOM.findDOMNode(this.validationMessage); // eslint-disable-line react/no-find-dom-node
+  }
+
+  /**
+   * Calculates various dimensions needed for render
+   *
+   * @method calculatePosition
+   * @return {Object}
+   */
+  calculatePosition = (icon, message, input) => {
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: can we not get away with just using the icon and message (the icon should give position etc)? or is the input essential?

if we don't need the input then we can get rid of the additional ref right?

File: src/utils/decorators/input-validation/input-validation.js [hunk 5]
```diff
@@ -627,31 +692,34 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
...
+          ref={(node) => { this._portal = node; }}
         >
-          { this.state.errorMessage || this.state.warningMessage || this.state.infoMessage }
+          <Icon
+            key='0'
+            ref={ (validationIcon) => { this.validationIcon = validationIcon; } }
+            type={ type }
+            className={ iconClasses }
+            style={ iconStyle }
+          />
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: does the icon need to be fed through the portal? can't it just render in position? (maybe this is why you need the input ref?)

File: src/utils/decorators/input-validation/input-validation.scss [hunk 6]
```diff
@@ -0,0 +1,7 @@
+.common-input__validation {
+  .common-input__message {
+    opacity: 0;
+    transition: opacity 0.2s 1.3s, visibility 0s 1.5s;
+    visibility: hidden;
+  }
+}
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.scss]: whats this used for?

**Comment by whomwah** [comment]
@andrewjtait I have attempted to address your various comments including memoizing. Now working on specs.

File: src/utils/decorators/input-validation/input-validation.js [hunk 7]
```diff
@@ -27,10 +29,11 @@ import chainFunctions from './../../helpers/chain-functions';
  *
  * In the render method for your component, you can now output the HTML:
  *
+ *   NOTE The targetted JSX must also have a ref of _input
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: is this still required? there is no reference to it in your changes

File: src/utils/decorators/input-validation/input-validation.js [hunk 8]
```diff
@@ -27,10 +29,11 @@ import chainFunctions from './../../helpers/chain-functions';
...
+ *
  *   render() {
  *     return (
  *       <div>
- *         <input />
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: may as well leave this in, it shows the order they should be rendered

File: src/utils/decorators/input-validation/input-validation.js [hunk 9]
```diff
@@ -227,6 +335,20 @@ const InputValidation = ComposedComponent => class Component extends ComposedCom
   }
 
   /**
+   * A lifecycle method for when the component has mounted.
+   *
+   * @method componentDidMount
+   * @return {void}
+   */
+  componentDidMount() {
```
Comment (andrewjtait) [inline src/utils/decorators/input-validation/input-validation.js]: does this need to sit alongside the other lifecycle methods nearer the top?

File: src/components/date/date.js [hunk 10]
```diff
@@ -180,6 +180,21 @@ class Date extends React.Component {
...
 
   /**
+   * A lifecycle method for when the component is removed from the page.
+   *
+   * @method componentWillUnmount
+   * @return {void}
+   */
+  componentWillUnmount() {
+    /* istanbul ignore if */
+    if (super.componentWillUnmount) { super.componentWillUnmount(); }
```
Comment (SeanArmstrong) [inline src/components/date/date.js]: Sorry if this has been discussed before but when wouldn't a parent have a componentWillUnmount?

Is this an issue with tests

Comment (whomwah) [reply src/components/date/date.js]: Fair point @SeanArmstrong. TBH I was just copying the idiom from the other lifecycle functions as I'd not seen it before. Could the same be said of the other conditional calls in the file to the `super` lifecycle functions?

Comment (whomwah) [reply src/components/date/date.js]: @SeanArmstrong this is my mistake. I think I saw this conditional in the input validation Decorator which probably needs it as it doesn't inherit from `React.Component`. I had to add these two lifecycle functions to date, and have not removed the calls to `super`. I'll push an update.

**Comment by andrewjtait** [comment]
Trying to think about how this reposition logic can be reused across other components, as we're going to need something similar for everything else using `ReactPortal`.

Do you think it should actually be our `ReactPortal` component which determines when the reposition should occur? It could receive a prop such as `onReposition` or something which you can pass the function to, it works out when it needs to be called?

**Comment by whomwah** [comment]
@andrewjtait are you assuming that the `ReactPortal` would only be used for setting up the various event listeners then? and then calling back to actually apply the positioning. For `resize` the event listener is applied to the window, but for `scroll` there was `context` in order to determine where to listen for the scroll events. Although I guess this is currently only ever going to be inside a Modal.

**Comment by andrewjtait** [comment]
Yeah the ReactPortal could set up the listeners, it could use context to look for a modal.

I can't think of a way we could do this without using context unfortunately?

**Comment by ianoxley** [comment]
Would the [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) be useful here?

There's a [polyfill](https://github.com/w3c/IntersectionObserver/tree/gh-pages/polyfill) available too for Safari and IE.

**Comment by whomwah** [comment]
@ianoxley @andrewjtait TBH I'd not seen this before and it's looks super useful. In this use case though for example we need to redraw if the scroll area moves even a pixel or two up or down so the label follows the input. Maybe I'm not understanding but I can't see where the intersection the `Intersection Observer` needs would take place?

**Pull request closed without merging**